### PR TITLE
Trainer Service and Cloud Functions cleanup

### PR DESCRIPTION
**Pull request by harrykeightley**
This is a combination of two large commits which really should have been done in isolation. 

### Trainer
1. Created a basic Trainer service! This is a cloud run service that gets triggered when model training jobs come in.
2. Overall, heavily simplified from the desktop version with later huggingface features.
3. Unit tests for non-flask parts of the service.
4. Little integration test! Feels very weird to train a model completely through pytest.

### Cloud functions
1. Move cloud functions to `/functions` folder from `/api/functions`. Removing one unnecessary layer of nesting (and also, they werent an api?)
5. Setup a poetry dev environment for the functions and upgrade their dependencies to versions compatible with python 3.10.
6. Create an Annotation dataclass at `fuctions/utils/annotation.py` to model annotations which were previously a dictionary.
7. Refactor the old `elan_to_json.py` to 3 functions which generate Annotations based on the requested tier selector.
8. Create tests for the above 2 points
9. Begin to use Annotations to make typing clearer throughout `functions/functions/datasets/process_file.py` cloud function.
10. Begin rewriting print statements with loguru.
11. Add new cloud function to process a model, adding a 'training job' to a pubsub topic.

### CODE REVIEW

**Review by nicklambourne — CHANGES_REQUESTED** [review]
Looks great 👍. A few comments here and there. Would appreciate smaller and more frequent PRs in the future.

File: architecture/modules/functions/variables.tf [hunk 1]
```diff
@@ -9,4 +9,5 @@ variable "elpis_worker" {}
 variable "user_upload_files_bucket" {}
 variable "user_datasets_bucket" {}
 
-variable "dataset_processing_topic" {}
\ No newline at end of file
+variable "dataset_processing_topic" {}
```
[the remark this answers is not in the document]
Comment (harrykeightley) [reply architecture/modules/functions/variables.tf:12]: The comments you left on the terraform side of things made me realise it's probably worth taking a harder look at the structure of the architecture folder, adding types and documentation and formatting, so I've created another branch to do so here: https://github.com/CoEDL/elpiscloud/tree/architecture_improvements, in which I'll address these comments.

File: functions/pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,28 @@
+[tool.poetry]
+name = "functions"
+version = "0.1.0"
+description = ""
+authors = ["Harry Keightley <harrykeightley@outlook.com>"]
```
[the remark this answers is not in the document]
Comment (benfoley) [reply functions/pyproject.toml:5]: But this way ppl can spam Harry on his outlook mail account into perpetuity. UQ address won't persist :(

File: architecture/modules/functions/variables.tf (hunk 1 above)
[the remark this answers is not in the document]
Comment (harrykeightley) [reply architecture/modules/functions/variables.tf:12]: Sadly, the variables weren't typed because I wasn't too familiar with terraform and that was what they had in the tutorials I was working through. With something like a third party pubsub topic, can I still use that as a type inside `variables.tf`?  Or should I be leaving  the more complex types as any?

File: services/trainer/trainer/trainer.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
+from dataclasses import dataclass
```
[the remark this answers is not in the document]
Comment (benfoley) [reply services/trainer/trainer/trainer.py:1]: Triggered! This reminds me of `/elpis/elpis/` 😅

File: architecture/modules/service/main.tf [hunk 4]
```diff
@@ -0,0 +1,52 @@
+resource "google_cloud_run_service" "default" {
+    name     = var.service_name
+    location = var.location
+    template {
+      spec {
+            containers {
```
[the remark this answers is not in the document]
Comment (harrykeightley) [reply architecture/modules/service/main.tf:6]: Fixed in https://github.com/CoEDL/elpiscloud/tree/architecture_improvements, can't believe terraform came with a formatter and I didn't even check.

File: architecture/modules/service/main.tf [hunk 5]
```diff
@@ -0,0 +1,52 @@
+resource "google_cloud_run_service" "default" {
```
[the remark this answers is not in the document]
Comment (harrykeightley) [reply architecture/modules/service/main.tf:1]: We absolutely should have all the cloud architecture documented. Kind of slack here not doing things incrementally, but my plan is (I can already see you shaking your head btw) push an MVP out for this and the transcriber, and then focus intensely on documentation and  increasing coverage. 

Here I was thinking that this module would be used for both the trainer and transcriber, since they'll both take job requests from a pubsub queue. This + the reason that the cloud run service is namespaced within the module made me rename this from 'trainer' back to 'default' since I thought there will only be one cloud run service inside here.

File: functions/functions/datasets/process_file.py [hunk 6]
```diff
@@ -1,39 +1,35 @@
...
-    """Background Cloud Function which triggers from pubsub dataset-processing
+def process_dataset_file(event, context) -> None:
+    """CloudEvent Function which triggers from pubsub dataset-processing
     topic.
 
     This processes the incoming file, whose path it can infer from the event. It
     cleans the data inside if it's a transcription file, and saves it to cloud
     storage so that it can be used in training with the dataset it came from.
 
     Parameters:
```
[the remark this answers is not in the document]
Comment (harrykeightley) [reply functions/functions/datasets/process_file.py:31]: Are we aiming in particular for Google-style docstrings? At this stage I've used Parameters for every docstring I've written in the project, is there a benefit to changing them over?

File: functions/requirements.txt [hunk 7]
```diff
@@ -0,0 +1,52 @@
+cachecontrol==0.12.12; python_version >= "3.6"
```
[the remark this answers is not in the document]
Comment (harrykeightley) [reply functions/requirements.txt:1]: I've just added a README within the `functions` folder describing this temporary process.
Pretty much cloud functions need a requirements.txt file for gcp to know what to do with them, and in the meantime I thought it could just be manually generated until this part is integrated into the pipeline before terraform runs.

File: services/trainer/trainer/trainer.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (harrykeightley) [reply services/trainer/trainer/trainer.py:1]: This is because of the nested poetry structure:
- The outer trainer is the entire folder for the service
- The inner trainer fikder is the python package name
- The trainer file is where the training logic takes place within the package
It does look gross though- what do you recommend?

File: functions/utils/model_metadata.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
+from dataclasses import dataclass, fields
+from enum import Enum
+from typing import Any, Dict
+
+from humps.main import decamelize
+
+from .firestore_event_converter import unpack
```
[the remark this answers is not in the document]
Comment (harrykeightley) [reply functions/utils/model_metadata.py]: Fixed in latest revision.

File: services/trainer/trainer/dataset.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+from functools import reduce
+from pathlib import Path
+from typing import Dict, List
+
+from datasets import Audio, load_dataset
+from datasets.dataset_dict import DatasetDict
+from trainer.model_metadata import ModelMetadata
+from transformers.models.wav2vec2.processing_wav2vec2 import Wav2Vec2Processor
+
+from .utterance import Utterance
```
[the remark this answers is not in the document]
Comment (harrykeightley) [reply services/trainer/trainer/dataset.py]: Fixed in latest revision.

File: services/trainer/trainer/dataset.py [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+) -> DatasetDict:
+    processed_path = dataset_path / "processed"
+    _process_dataset(dataset_path, processed_path)
+
+    dataset = load_dataset(
+        "json", cache_dir=str(cache_dir), data_dir=str(processed_path)
+    )
+    dataset = dataset.cast_column("audio", Audio(sampling_rate=metadata.sampling_rate))
+
+    return dataset["train"].train_test_split(test_size=0.2)  # type: ignore
```
[the remark this answers is not in the document]
Comment (harrykeightley) [reply services/trainer/trainer/dataset.py]: Fixed in latest revision.

File: services/trainer/trainer/dataset.py [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+        with open(file) as f:
+            utterances = map(Utterance.from_dict, json.load(f))
+        utterance = reduce(Utterance.combine, utterances)
+
+        path = output_dir / file.name
+        with open(path, "w") as out_file:
+            json.dump(utterance.to_dict(dataset_dir), out_file)
+
+
+def prepare_dataset(dataset: DatasetDict, processor: Wav2Vec2Processor) -> DatasetDict:
```
[the remark this answers is not in the document]
Comment (harrykeightley) [reply services/trainer/trainer/dataset.py:75]: Fixed in latest revision.

File: services/trainer/trainer/utterance.py [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+    transcript: str
+    start_ms: Optional[int]
+    stop_ms: Optional[int]
+    speaker_id: Optional[str]
+
+    def combine(self, utterance: "Utterance") -> "Utterance":
+        return Utterance(
+            audio_file_name=self.audio_file_name,
+            transcript=f"{self.transcript} {utterance.transcript}",
+            start_ms=self.start_ms,
```
[the remark this answers is not in the document]
Comment (harrykeightley) [reply services/trainer/trainer/utterance.py]: In the next revision, I have just removed utterances having timestamps at all. 

The plan is (in preprocessing in the cloud functions) to split the audio files/transcripts into utterances so that each utterance has its own audio file of which it spans the entire length.

File: services/trainer/trainer/utterance.py [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+@dataclass
+class Utterance:
+    audio_file_name: str
+    transcript: str
+    start_ms: Optional[int]
+    stop_ms: Optional[int]
+    speaker_id: Optional[str]
+
+    def combine(self, utterance: "Utterance") -> "Utterance":
+        return Utterance(
```
[the remark this answers is not in the document]
Comment (harrykeightley) [reply services/trainer/trainer/utterance.py:25]: Fixed in latest revision.

File: services/trainer/trainer/trainer.py [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+
+import torch
+from loguru import logger
+from transformers import AutoModelForCTC, AutoProcessor, Trainer
+
+from .dataset import create_dataset, prepare_dataset
+from .model_metadata import ModelMetadata
+
+
+def train(
```
[the remark this answers is not in the document]
Comment (harrykeightley) [reply services/trainer/trainer/trainer.py]: Fixed in latest revision.

File: services/trainer/trainer/main.py [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+    model_path = train(
+        metadata=metadata, data_path=DATA_PATH, dataset_path=dataset_path
+    )
+    if model_path is None:
+        msg = "Training failed, exiting without uploading model"
+        logger.error(f"Error: {msg}")
+        return f"Error: {msg}", 500
+
+    upload_results(metadata, model_path)
+    return ("", 204)
```
[the remark this answers is not in the document]
Comment (harrykeightley) [reply services/trainer/trainer/main.py:63]: Yes 👍 I'm never using the response and just want to know that the task completed.

File: functions/utils/annotation.py [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+    Annotation spans the entire audio file.
+    """
+
+    audio_file_name: str
+    transcript: str
+    start_ms: Optional[int] = None
+    stop_ms: Optional[int] = None
+    speaker_id: Optional[str] = None
+
+    def rescale_times(self, old_sample_rate: int, sample_rate: int) -> "Annotation":
```
[the remark this answers is not in the document]
Comment (harrykeightley) [reply functions/utils/annotation.py:24]: I tossed this up for a bit before I made the initial commit- I think the best thing is to include the sample rate as part of the annotation, but I will do this in a later PR where I'm updating the `process_file` cloud function where it's used.

File: functions/functions/datasets/process_file.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (nicklambourne) [reply functions/functions/datasets/process_file.py:31]: I don't mind which style we use, but the advantage to picking a commonly supported one and using it consistently throughout means that you can use autodocumentation generators like Sphinx and have everything come out as expected, something I expect we'll eventually want to do.

File: architecture/modules/service/main.tf (hunk 5 above)
[the remark this answers is not in the document]
Comment (nicklambourne) [reply architecture/modules/service/main.tf:1]: > Here I was thinking that this module would be used for both the trainer and transcriber, since they'll both take job requests from a pubsub queue. This + the reason that the cloud run service is namespaced within the module made me rename this from 'trainer' back to 'default' since I thought there will only be one cloud run service inside here.

Sounds fair enough to me.

> We absolutely should have all the cloud architecture documented. Kind of slack here not doing things incrementally, but my plan is (I can already see you shaking your head btw) push an MVP out for this and the transcriber, and then focus intensely on documentation and increasing coverage.

MVPs are fine, but you do have to make sure that you actually come back, sooner rather than later. The bus factor on this project is small and I'm sure neither Ben nor I want to have to understand this from scratch if that bus (or job offer, or whatever) ever did come.

File: architecture/modules/functions/variables.tf (hunk 1 above)
[the remark this answers is not in the document]
Comment (nicklambourne) [reply architecture/modules/functions/variables.tf:12]: No worries. I'd honestly just suggest inlining all variables in the module files, there's really no reason for separate files. Define the types as far as practical and consider using outputs for the key values (ids, names, urls etc.) that you need to minimise the need for complex types, but if you still do need them, something like:
```
variable "stages" {
  type = map(object({
    storage_integration = string,
    url = string,
  }))
}
```
Is perfectly valid.

File: services/trainer/trainer/trainer.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (nicklambourne) [reply services/trainer/trainer/trainer.py:1]: I'd probably aim for something like `services/trainer/src/trainer.py`. Because of the way poetry works you should be able reference the module elsewhere with `import trainer.src.trainer as trainer; trainer.blah()` outside the trainer file if you need to.

File: services/trainer/README.md [hunk 17]
```diff
@@ -0,0 +1,15 @@
...
+transcription jobs.
+
+## Installation
+
+- Make sure you have poetry installed, and a compatible python version.
+- Create a virtual env and install the dependencies: `poetry install`
+
+## Running the tests
+
+- Unit tests: `poetry run pytest -m 'not integration'`
```
Comment (nicklambourne) [inline services/trainer/README.md:14]: `'not integration'`? 😂

**Review by nicklambourne — APPROVED** [review]
LGTM w/ a few comments/responses.

File: services/trainer/README.md (hunk 17 above)
Comment (harrykeightley) [reply services/trainer/README.md:14]: Runs all the tests except those marked as integration tests :) 
I guess I could also mark all the unit tests as unit but that seems boilerplatey
